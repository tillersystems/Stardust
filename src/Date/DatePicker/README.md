# DatePicker

### Usage

```jsx
import DatePicker from 'components/Date/DatePicker';

// or

import { DatePicker } from 'components';
```

<!-- STORY -->

### Properties

- `className` - Class needed by styled component.
- `locale` - The locale to display the weekday names.
- `value` - The value of the selected date (controlled mode).
- `defaultValue` - The default selected date (uncontrolled mode).
- `minDate` - The minimum selectable date.
- `maxDate` - The maximum selectable date.
- `onChange` - Handler on date changed.

| `propName`     |     propType     | defaultValue | isRequired |
| -------------- | :--------------: | :----------: | :--------: |
| `className`    |     `string`     |     `''`     |     -      |
| `locale`       |     `string`     |     `en`     |     -      |
| `value`        | `luxon.DateTime` |    `null`    |     -      |
| `defaultValue` | `luxon.DateTime` |   `today`    |     -      |
| `minDate`      | `luxon.DateTime` |    `null`    |     -      |
| `maxDate`      | `luxon.DateTime` |    `null`    |     -      |
| `onChange`     |    `function`    |  `() => {}`  |     -      |

### Examples

The `DatePicker` can be used in two modes, in either controlled or uncontrolled mode.
In uncontrolled mode, the `DatePicker` manages the selected date internally (yet, the `onChange`
handler is called upon changes). In controlled mode, the selected date is based on the given `value`
prop. The `onChange` handler is called upon user changes, but the selected value won't be affected
unless the `value` prop is updated.

#### Uncontrolled example

```jsx
import { DatePicker } from 'components';

render() {
  return <DatePicker />
}
```

#### Controlled example

```jsx
import { DatePicker } from 'components';
import { DateTime } from 'luxon';

let value = DateTime.locale();

handleChange = date => value = date;

render() {
  return <DatePicker value={value} onChange={handleChange} />
}
```
