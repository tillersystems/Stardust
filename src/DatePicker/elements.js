import styled from 'styled-components';

export const Container = styled.div`
  &:not(:last-child) {
    margin-right: ${({ numberOfMonthsToDisplay }) =>
      numberOfMonthsToDisplay === 2 ? '2.4rem' : 0};
  }
`;
