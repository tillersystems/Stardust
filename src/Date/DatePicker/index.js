import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { DateTime } from 'luxon';

import { Button, Icon, Theme } from '../..';
import Weeks from '../Weeks';
import WeekdayNames from '../WeekdayNames';
import { Header, HeaderTitle } from '../elements';

/**
 * Date Picker
 *
 * This component is in charge of displaying a date picker.
 *
 * @param {string} className - Styled component class name.
 * @param {string} locale - Local to use to display the weekday names.
 * @param {luxon.DateTime} value - Controlled value (the selected date).
 * @param {luxon.DateTime} defaultValue - Default date (the default date).
 * @param {luxon.DateTime} minDate - Minimum allowed date.
 * @param {luxon.DateTime} maxDate - Maximum allowed date.
 * @param {function} onChange - Handler of date change.
 *
 * @return {jsx}
 */
class DatePicker extends PureComponent {
  /** Display name. */
  static displayName = 'DatePicker';

  /** Validation prop types. */
  static propTypes = {
    className: PropTypes.string,
    locale: PropTypes.string,
    value: PropTypes.object,
    defaultValue: PropTypes.object,
    minDate: PropTypes.object,
    maxDate: PropTypes.object,
    onChange: PropTypes.func,
  };

  /** Default props. */
  static defaultProps = {
    className: '',
    locale: 'en',
    value: null,
    defaultValue: DateTime.local(),
    minDate: null,
    maxDate: null,
    onChange: () => {},
  };

  /** Internal state. */
  state = {
    currentMonth: DateTime.local(),
    selectedDate: null,
    highlightedDate: null,
  };

  /**
   * Handles date click.
   *
   * @param {luxon.DateTime} date - The date that was clicked.
   */
  handleDateClick = date => {
    const { onChange } = this.props;
    this.setState({ ...this.state, selectedDate: date, currentMonth: date }, () => {
      onChange(date);
    });
  };

  /**
   * Handles mouse over a date.
   *
   * @param {luxon.DateTime} date - The date the mouse is over.
   */
  handleDateOver = date => {
    this.setState({ ...this.state, highlightedDate: date });
  };

  /**
   * Handles click on the previous month button.
   */
  handlePrevMonthClick = () => {
    const { currentMonth } = this.state;
    this.setState({ ...this.state, currentMonth: currentMonth.minus({ months: 1 }) });
  };

  /**
   * Handles click on the next month button.
   */
  handleNextMonthClick = () => {
    const { currentMonth } = this.state;
    this.setState({ ...this.state, currentMonth: currentMonth.plus({ months: 1 }) });
  };

  render() {
    const { className, locale, value, defaultValue, minDate, maxDate } = this.props;
    const { currentMonth, selectedDate, highlightedDate } = this.state;

    const actualSelectedDate = value || selectedDate || defaultValue;

    const isAfterMinDate = !minDate || currentMonth.startOf('month') >= minDate.startOf('month');
    const isBeforeMaxDate = !maxDate || currentMonth.endOf('month') <= maxDate.endOf('month');

    return (
      <div className={className} onMouseLeave={() => this.handleDateOver(null)}>
        <Header>
          <Button
            data-testid="pervious-month-button"
            appearance="secondary"
            size="small"
            onClick={this.handlePrevMonthClick}
            disabled={!isAfterMinDate}
          >
            <Icon color={Theme.palette.darkBlue} name="chevron-left" />
          </Button>
          <HeaderTitle>
            {currentMonth.setLocale(locale).toLocaleString({ month: 'long', year: 'numeric' })}
          </HeaderTitle>
          <Button
            data-testid="next-month-button"
            appearance="secondary"
            size="small"
            onClick={this.handleNextMonthClick}
            disabled={!isBeforeMaxDate}
          >
            <Icon color={Theme.palette.darkBlue} name="chevron-right" />
          </Button>
        </Header>
        <>
          <WeekdayNames currentMonth={currentMonth} locale={locale} />
          <Weeks
            currentMonth={currentMonth}
            selectedStartDate={actualSelectedDate ? actualSelectedDate.startOf('day') : null}
            selectedEndDate={actualSelectedDate ? actualSelectedDate.endOf('day') : null}
            highlightedStartDate={highlightedDate ? highlightedDate.startOf('day') : null}
            highlightedEndDate={highlightedDate ? highlightedDate.endOf('day') : null}
            minDate={minDate}
            maxDate={maxDate}
            onDateClick={this.handleDateClick}
            onDateOver={this.handleDateOver}
          />
        </>
      </div>
    );
  }
}

export default styled(DatePicker)`
  width: 20rem;
`;
