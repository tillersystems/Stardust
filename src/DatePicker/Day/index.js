import React from 'react';
import PropTypes from 'prop-types';

import { Container, Content } from './elements';

/**
 * A single Day component.
 *
 * @param {node} children -
 * @param {node} disabled -
 * @param {boolean} isSelected - Whether the day is selected or not.
 * @param {boolean} isStartEdge -
 * @param {boolean} isEndEdge -
 * @param {boolean} isInPath -
 * @param {boolean} shadowed - Whether the day is shadowed or not.
 * @param {function} onClick - Handler called when the day is clicked.
 *
 * @return {jsx}
 */
const Day = ({
  disabled,
  isSelected,
  shadowed,
  onClick,
  children,
  isInPath,
  isStartEdge,
  isEndEdge,
}) => (
  <Container onClick={onClick} isInPath={isInPath} isStartEdge={isStartEdge} isEndEdge={isEndEdge}>
    <Content
      disabled={disabled}
      isSelected={isSelected}
      isStartEdge={isStartEdge}
      isEndEdge={isEndEdge}
      shadowed={shadowed}
    >
      {children}
    </Content>
  </Container>
);

/**
 * PropTypes Validation.
 */
Day.displayName = 'Day';

/**
 * Default props.
 */
const { bool, func, node } = PropTypes;
Day.propTypes = {
  disabled: bool,
  isSelected: bool,
  isStartEdge: bool,
  isEndEdge: bool,
  isInPath: bool,
  shadowed: bool,
  onClick: func,
  children: node.isRequired,
};

/**
 * Default props.
 */
Day.defaultProps = {
  disabled: false,
  isSelected: false,
  isStartEdge: false,
  isEndEdge: false,
  isInPath: false,
  shadowed: false,
  onClick: () => {},
};

export default Day;
