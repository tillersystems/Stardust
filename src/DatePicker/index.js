import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { DateTime, Interval } from 'luxon';

import Weeks from './Weeks';
import Weekdays from './Weekdays';
import Header from './Header';
import { Container } from './elements';
import { getPreviousMonth, getNextMonth, isSameDay, isInNextMonth, isInLastMonth } from './helpers';

const { bool, func, number, object, string } = PropTypes;
const localDateTime = DateTime.local();

/**
 * Date Picker
 *
 * This component is in charge of displaying a date picker.
 *
 * @param {string} className - Styled component class name.
 * @param {string} locale - Local to use to display the weekday names.
 * @param {luxon.DateTime} value - Controlled value (the selected date).
 * @param {luxon.DateTime} minDate - Minimum allowed date.
 * @param {luxon.DateTime} maxDate - Maximum allowed date.
 * @param {number} numberOfMonthsToDisplay - The number of month to display (1 or 2).
 * @param {function} onDateChanged - Handler of date change.
 * @param {bool} rangePicker - Whether it is a date picker or a date range picker.
 *
 * @return {jsx}
 */
class DatePicker extends PureComponent {
  /** Display name. */
  static displayName = 'DatePicker';

  /** Validation prop types. */
  static propTypes = {
    className: string,
    locale: string,
    value: object,
    minDate: object,
    maxDate: object,
    numberOfMonthsToDisplay: number,
    onDateChanged: func,
    rangePicker: bool,
  };

  /** Default props. */
  static defaultProps = {
    className: '',
    locale: 'en',
    value: null,
    minDate: null,
    maxDate: null,
    numberOfMonthsToDisplay: 1,
    onDateChanged: () => {},
    rangePicker: false,
  };

  /** Internal state. */
  state = {
    /** Initial Calendar date */
    currentDate: localDateTime,

    /** Selected Calendar date */
    selected: localDateTime,

    /** Start date value */
    startDate: null,

    /** End date value */
    endDate: null,
  };

  /**
   * Component lifecycle method
   * Update current month to keep track of the date if provided
   *
   */
  componentDidMount() {
    const { value } = this.props;

    // Set selected & currentDate in componentDidMount if is relies on props
    if (value) {
      this.setState({ selected: value, currentDate: value });
    }
  }

  /**
   * Component lifecycle method
   *
   */
  componentDidUpdate(prevProps) {
    const { value: prevValue } = prevProps;
    const { value } = this.props;

    if (value !== prevValue) {
      this.setState({ selected: value, currentDate: value });
    }
  }

  /**
   * Get selected Range
   * Take the date range selection
   * and return as an interval
   *
   * @return {luxon.Interval}
   */
  getSelectedRange() {
    const { startDate, endDate } = this.state;
    const { onDateChanged } = this.props;
    const selectedRange = !!startDate && !!endDate;

    if (!selectedRange) {
      return;
    }

    return onDateChanged(Interval.fromDateTimes(startDate, endDate));
  }

  /**
   * Get selected
   * Take the selected date
   *
   * @return {luxon.DateTime}
   */
  getSelected() {
    const { selected } = this.state;
    const { onDateChanged } = this.props;

    return onDateChanged(selected);
  }

  /**
   * Handles date change.
   *
   * @param {luxon.DateTime} value - The date that was clicked.
   */
  handleChange = value => {
    const { rangePicker, numberOfMonthsToDisplay } = this.props;
    const { selected: previousValue, startDate, endDate, currentDate } = this.state;

    if (rangePicker) {
      let start = startDate;
      let end = endDate;

      if (!start) {
        start = value;
      } else {
        if (endDate) {
          start = value;
          end = undefined;
        } else {
          if (value >= start) {
            end = value;
          } else {
            end = start;
            start = value;
          }
        }
      }

      // We check whether value is in the next month (w.r.t. the whole calendar)
      // or the previous month
      if (
        isInNextMonth(value, currentDate, numberOfMonthsToDisplay) ||
        isInLastMonth(value, currentDate)
      ) {
        this.setState(
          { selected: value, startDate: start, endDate: end, currentDate: value },
          () => {
            this.getSelectedRange();
          },
        );
      } else {
        this.setState({ selected: value, startDate: start, endDate: end }, () => {
          this.getSelectedRange();
        });
      }
    } else {
      // First we check if the current value is different from the previsous one,
      // we never call a setState on a already selected value.
      // Then we check whether value is in the next month (w.r.t. the whole calendar)
      // or the previous month
      if (!isSameDay(value, previousValue)) {
        if (
          isInNextMonth(value, currentDate, numberOfMonthsToDisplay) ||
          isInLastMonth(value, currentDate)
        ) {
          this.setState({ selected: value, currentDate: value }, () => {
            this.getSelected();
          });
        } else {
          this.setState({ selected: value }, () => {
            this.getSelected();
          });
        }
      }
    }
  };

  /**
   * Handles click on the previous month button.
   */
  handlePrevMonthClick = () => {
    const { currentDate } = this.state;

    this.setState({ currentDate: getPreviousMonth(currentDate) });
  };

  /**
   * Handles click on the next month button.
   */
  handleNextMonthClick = () => {
    const { currentDate } = this.state;

    this.setState({ currentDate: getNextMonth(currentDate) });
  };

  render() {
    const { className, locale, minDate, maxDate, numberOfMonthsToDisplay } = this.props;
    const { currentDate, selected, startDate, endDate } = this.state;

    const isAfterMinDate = !minDate || currentDate.startOf('month') > minDate.startOf('month');
    const isBeforeMaxDate = !maxDate || currentDate.endOf('month') < maxDate.endOf('month');

    return (
      <div className={className} style={{ display: 'flex' }}>
        {[...Array(numberOfMonthsToDisplay).keys()].map(monthIndex => {
          const stringifiedDate = currentDate
            .plus({ month: monthIndex })
            .setLocale(locale)
            .toLocaleString({ month: 'long', year: 'numeric' });
          return (
            <Container key={monthIndex} numberOfMonthsToDisplay={numberOfMonthsToDisplay}>
              <Header
                title={stringifiedDate}
                numberOfMonthsToDisplay={numberOfMonthsToDisplay}
                monthIndex={monthIndex}
                prev={this.handlePrevMonthClick}
                next={this.handleNextMonthClick}
                shouldDisablePrev={!isAfterMinDate}
                shouldDisableNext={!isBeforeMaxDate}
              />
              <Weekdays locale={locale} />
              <Weeks
                currentDate={currentDate.plus({ month: monthIndex })}
                selected={selected}
                minDate={minDate}
                maxDate={maxDate}
                onDateClick={this.handleChange}
                startDate={startDate}
                endDate={endDate}
              />
            </Container>
          );
        })}
      </div>
    );
  }
}

export default styled(DatePicker)`
  width: ${({ numberOfMonthsToDisplay }) =>
    numberOfMonthsToDisplay === 2 ? 'calc((24.4rem * 2) + 2.4rem)' : '24.4rem'};
  font-weight: ${({ theme: { fonts } }) => fonts.weight.thick};
`;
