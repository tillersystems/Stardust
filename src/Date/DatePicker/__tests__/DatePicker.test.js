import React from 'react';
import * as Luxon from 'luxon';
import { fireEvent } from 'react-testing-library';

import DatePicker from '..';

describe('<DatePicker />', () => {
  const mockLocalDateTime = Luxon.DateTime.fromObject({
    year: 2018,
    month: 7,
    day: 15,
    hour: 0,
    minute: 0,
    second: 0,
    zone: 'Europe/Paris',
  });

  beforeEach(() => {
    Luxon.DateTime.local = jest.fn().mockImplementation(() => mockLocalDateTime);
  });

  test('should render without a problem', () => {
    const { container } = render(
      <DatePicker minDate={mockLocalDateTime} defaultValue={mockLocalDateTime} />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should select a date', () => {
    const { getByText } = render(<DatePicker defaultValue={mockLocalDateTime} />);

    const daySelected = getByText('20');

    // Click on previous month button
    fireEvent.click(daySelected);

    expect(daySelected).toBeInTheDocument();
  });

  test('should display the new provided date', () => {
    const { rerender, getByText } = render(<DatePicker value={mockLocalDateTime} />);
    const daySelected = getByText('15');

    expect(daySelected).toHaveStyleRule('border-top-left-radius', '0.4rem');
    expect(daySelected).toHaveStyleRule('border-top-right-radius', '0.4rem');
    expect(daySelected).toBeInTheDocument();

    const newDate = Luxon.DateTime.fromObject({
      year: 2018,
      month: 9,
      day: 20,
      hour: 0,
      minute: 0,
      second: 0,
      zone: 'Europe/Paris',
    });
    rerender(<DatePicker value={newDate} />);

    const newDaySelected = getByText('20');

    expect(newDaySelected).toHaveStyleRule('border-top-left-radius', '0.4rem');
    expect(newDaySelected).toHaveStyleRule('border-top-right-radius', '0.4rem');
    expect(newDaySelected).toBeInTheDocument();
  });

  test('should handle previous month', () => {
    const { getByTestId, getByText } = render(<DatePicker defaultValue={mockLocalDateTime} />);

    const previousMonthButton = getByTestId('previous-month-button');

    // Click on previous month button
    fireEvent.click(previousMonthButton);

    const excpectedMonth = getByText(/June/);

    expect(excpectedMonth).toBeInTheDocument();
  });

  test('should handle next month', () => {
    const { getByTestId, getByText } = render(<DatePicker defaultValue={mockLocalDateTime} />);

    const nextMonthButton = getByTestId('next-month-button');

    // Click on previous month button
    fireEvent.click(nextMonthButton);

    const excpectedMonth = getByText(/August/);

    expect(excpectedMonth).toBeInTheDocument();
  });

  test('should handle mouse Leave', () => {
    const { container } = render(<DatePicker defaultValue={mockLocalDateTime} />);

    // Mouse Leave on DatePicker
    fireEvent.mouseLeave(container.firstChild);

    expect(container.firstChild).toBeInTheDocument();
  });
});
