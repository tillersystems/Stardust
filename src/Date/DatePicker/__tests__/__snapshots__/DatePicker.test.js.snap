// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should render without a problem 1`] = `
.c2 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.8rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(207,13%,45%);
  background: linear-gradient(180deg,hsl(0,100%,100%) 0%,hsl(216,45%,98%) 100%);
  border: 1px solid hsl(207,22%,90%);
  padding: 0.35rem 1.4rem;
}

.c2 svg > path {
  fill: hsl(0,100%,100%);
}

.c2 svg > path {
  fill: hsl(207,13%,45%);
}

.c8 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  cursor: not-allowed;
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
}

.c8 > .c9 {
  color: hsl(207,22%,90%);
}

.c8 > .c9 {
  color: hsl(207,22%,90%);
}

.c11 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  cursor: not-allowed;
}

.c11 > .c9 {
  color: hsl(207,22%,90%);
}

.c11 > .c9 {
  color: hsl(207,22%,90%);
}

.c13 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  cursor: not-allowed;
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c13 > .c9 {
  color: hsl(207,22%,90%);
}

.c15 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  cursor: not-allowed;
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
}

.c15 > .c9 {
  color: hsl(207,22%,90%);
}

.c16 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  cursor: not-allowed;
}

.c16 > .c9 {
  color: hsl(207,22%,90%);
}

.c17 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  cursor: not-allowed;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c17 > .c9 {
  color: hsl(207,22%,90%);
}

.c19 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c19 > .c9 {
  color: hsl(0,100%,100%);
  background: hsl(200,74%,46%);
}

.c20 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
}

.c21 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
}

.c22 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c23 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
}

.c23 > .c9 {
  color: hsl(207,22%,90%);
}

.c24 {
  font-size: 1.4rem;
  line-height: 3rem;
  cursor: pointer;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c24 > .c9 {
  color: hsl(207,22%,90%);
}

.c10 {
  width: 100%;
  height: 100%;
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
}

.c12 {
  width: 100%;
  height: 100%;
}

.c14 {
  width: 100%;
  height: 100%;
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c18 {
  width: 100%;
  height: 100%;
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

.c7 {
  display: grid;
  grid-template-columns: repeat(7,2.8rem);
  grid-template-rows: 2.8rem;
  justify-items: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  text-align: center;
  color: hsl(206,23%,69%);
}

.c6 {
  font-size: 1.4rem;
}

.c5 {
  display: grid;
  grid-template-columns: repeat(7,2.8rem);
  grid-template-rows: 2rem;
  justify-items: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  text-align: center;
  border-bottom: 1px solid hsl(206,23%,69%);
  color: hsl(206,23%,69%);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 1.5rem;
  color: hsl(213,17%,20%);
  font-size: 1.4rem;
  line-height: 1.4rem;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-align: center;
}

.c3 {
  display: inline-block;
  border-image-source: none;
  border-image-width: 0;
  padding: 0;
  font-size: 20;
  line-height: 20;
}

.c0 {
  width: 20rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <button
      class="c2"
      data-testid="previous-month-button"
      type="button"
    >
      <div
        class="c3"
        data-testid="iconContainer"
        height="20"
        width="20"
      >
        <svg
          data-testid="iconSvg"
          height="20"
          style="display: block;"
          viewBox="0 0 512 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M335.085 368.915c8.331 8.331 8.331 21.839 0 30.17-8.331 8.331-21.839 8.331-30.17 0l-128-128c-8.331-8.331-8.331-21.839 0-30.17l128-128c8.331-8.331 21.839-8.331 30.17 0 8.331 8.331 8.331 21.839 0 30.17L222.17 256l112.915 112.915z"
            data-testid="iconSvgPath"
            fill="hsl(213,17%,20%)"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </button>
    <div
      class="c4"
    >
      July 2018
    </div>
    <button
      class="c2"
      data-testid="next-month-button"
      type="button"
    >
      <div
        class="c3"
        data-testid="iconContainer"
        height="20"
        width="20"
      >
        <svg
          data-testid="iconSvg"
          height="20"
          style="display: block;"
          viewBox="0 0 512 512"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M176.915 143.085c-8.331-8.331-8.331-21.839 0-30.17 8.331-8.331 21.839-8.331 30.17 0l128 128c8.331 8.331 8.331 21.839 0 30.17l-128 128c-8.331 8.331-21.839 8.331-30.17 0-8.331-8.331-8.331-21.839 0-30.17L289.83 256 176.915 143.085z"
            data-testid="iconSvgPath"
            fill="hsl(213,17%,20%)"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    class="c5"
  >
    <div
      class="c6"
    >
      Mon
    </div>
    <div
      class="c6"
    >
      Tue
    </div>
    <div
      class="c6"
    >
      Wed
    </div>
    <div
      class="c6"
    >
      Thu
    </div>
    <div
      class="c6"
    >
      Fri
    </div>
    <div
      class="c6"
    >
      Sat
    </div>
    <div
      class="c6"
    >
      Sun
    </div>
  </div>
  <div>
    <div
      class="c7"
    >
      <div
        class="c8"
        disabled=""
      >
        <div
          class="c9 c10"
        >
          25
        </div>
      </div>
      <div
        class="c11"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          26
        </div>
      </div>
      <div
        class="c11"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          27
        </div>
      </div>
      <div
        class="c11"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          28
        </div>
      </div>
      <div
        class="c11"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          29
        </div>
      </div>
      <div
        class="c11"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          30
        </div>
      </div>
      <div
        class="c13"
        disabled=""
      >
        <div
          class="c9 c14"
        >
          01
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c15"
        disabled=""
      >
        <div
          class="c9 c10"
        >
          02
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          03
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          04
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          05
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          06
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          07
        </div>
      </div>
      <div
        class="c17"
        disabled=""
      >
        <div
          class="c9 c18"
        >
          08
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c15"
        disabled=""
      >
        <div
          class="c9 c10"
        >
          09
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          10
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          11
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          12
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          13
        </div>
      </div>
      <div
        class="c16"
        disabled=""
      >
        <div
          class="c9 c12"
        >
          14
        </div>
      </div>
      <div
        class="c19"
      >
        <div
          class="c9 c14"
        >
          15
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c20"
      >
        <div
          class="c9 c10"
        >
          16
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          17
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          18
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          19
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          20
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          21
        </div>
      </div>
      <div
        class="c22"
      >
        <div
          class="c9 c18"
        >
          22
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c20"
      >
        <div
          class="c9 c10"
        >
          23
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          24
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          25
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          26
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          27
        </div>
      </div>
      <div
        class="c21"
      >
        <div
          class="c9 c12"
        >
          28
        </div>
      </div>
      <div
        class="c22"
      >
        <div
          class="c9 c18"
        >
          29
        </div>
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c20"
      >
        <div
          class="c9 c10"
        >
          30
        </div>
      </div>
      <div
        class="c22"
      >
        <div
          class="c9 c18"
        >
          31
        </div>
      </div>
      <div
        class="c23"
      >
        <div
          class="c9 c12"
        >
          01
        </div>
      </div>
      <div
        class="c23"
      >
        <div
          class="c9 c12"
        >
          02
        </div>
      </div>
      <div
        class="c23"
      >
        <div
          class="c9 c12"
        >
          03
        </div>
      </div>
      <div
        class="c23"
      >
        <div
          class="c9 c12"
        >
          04
        </div>
      </div>
      <div
        class="c24"
      >
        <div
          class="c9 c18"
        >
          05
        </div>
      </div>
    </div>
  </div>
</div>
`;
